File: src/prototype/web/tests/unit/cut_attempt.test.ts
// Tests for the defender cut-attempt FSM.
// docs/design/state_machines_v1.md §4.2.

import { describe, expect, it } from "vitest";
import {
  CUT_TIMING,
  INITIAL_CUT_ATTEMPTS,
  pickCutTarget,
  tickCutAttempts,
  type CutTickInput,
} from "../../src/state/cut_attempt.js";
import type { HandFSM } from "../../src/state/hand_fsm.js";
import type { GripZone } from "../../src/input/intent.js";

function gripped(side: "L" | "R", zone: GripZone): HandFSM {
  return Object.freeze({
    side,
    state: "GRIPPED" as const,
    target: zone,
    stateEnteredMs: 0,
    reachDurationMs: 0,
    lastParriedZone: null,
    lastParriedAtMs: Number.NEGATIVE_INFINITY,
  });
}
function idle(side: "L" | "R"): HandFSM {
  return Object.freeze({
    side,
    state: "IDLE" as const,
    target: null,
    stateEnteredMs: 0,
    reachDurationMs: 0,
    lastParriedZone: null,
    lastParriedAtMs: Number.NEGATIVE_INFINITY,
  });
}

function inputs(over: Partial<CutTickInput> = {}): CutTickInput {
  return {
    nowMs: 0,
    leftCommit: null,
    rightCommit: null,
    attackerLeft: idle("L"),
    attackerRight: idle("R"),
    attackerTriggerL: 0,
    attackerTriggerR: 0,
    ...over,
  };
}

describe("pickCutTarget (§B.4.1)", () => {
  it("returns null when no attacker hand is GRIPPED", () => {
    expect(pickCutTarget({ x: 0, y: 0 }, idle("L"), idle("R"))).toBe(null);
  });

  it("returns the only GRIPPED hand regardless of RS", () => {
    const t = pickCutTarget({ x: 0.5, y: 0.5 }, gripped("L", "SLEEVE_R"), idle("R"));
    expect(t).toEqual({ side: "L", zone: "SLEEVE_R" });
  });

  it("prefers attacker L when RS points left", () => {
    const t = pickCutTarget(
      { x: -1, y: 0 },
      gripped("L", "SLEEVE_R"),
      gripped("R", "SLEEVE_L"),
    );
    expect(t?.side).toBe("L");
  });

  it("prefers attacker R when RS points right", () => {
    const t = pickCutTarget(
      { x: 1, y: 0 },
      gripped("L", "SLEEVE_R"),
      gripped("R", "SLEEVE_L"),
    );
    expect(t?.side).toBe("R");
  });
});

describe("tickCutAttempts lifecycle", () => {
  it("commits start IN_PROGRESS with a target", () => {
    const res = tickCutAttempts(INITIAL_CUT_ATTEMPTS, inputs({
      nowMs: 0,
      leftCommit: { rs: { x: 0, y: 0 } },
      attackerLeft: gripped("L", "SLEEVE_R"),
    }));
    expect(res.next.left.kind).toBe("IN_PROGRESS");
    expect(res.events.some((e) => e.kind === "CUT_STARTED")).toBe(true);
  });

  it("silent drop when no GRIPPED hand available", () => {
    const res = tickCutAttempts(INITIAL_CUT_ATTEMPTS, inputs({
      leftCommit: { rs: { x: 1, y: 0 } },
    }));
    expect(res.next.left.kind).toBe("IDLE");
    expect(res.events).toEqual([]);
  });

  it("resolution SUCCEEDS when attacker trigger < 0.5 at expiry", () => {
    let s = INITIAL_CUT_ATTEMPTS;
    s = tickCutAttempts(s, inputs({
      nowMs: 0,
      leftCommit: { rs: { x: 0, y: 0 } },
      attackerLeft: gripped("L", "SLEEVE_R"),
      attackerTriggerL: 0.8,
    })).next;
    const res = tickCutAttempts(s, inputs({
      nowMs: CUT_TIMING.attemptMs,
      attackerLeft: gripped("L", "SLEEVE_R"),
      attackerTriggerL: 0.3, // attacker let go during attempt
    }));
    expect(res.events.some((e) => e.kind === "CUT_SUCCEEDED" && "attackerSide" in e && e.attackerSide === "L")).toBe(true);
    expect(res.next.left.kind).toBe("IDLE");
  });

  it("resolution FAILS when attacker trigger ≥ 0.5 at expiry", () => {
    let s = INITIAL_CUT_ATTEMPTS;
    s = tickCutAttempts(s, inputs({
      nowMs: 0,
      rightCommit: { rs: { x: 0, y: 0 } },
      attackerRight: gripped("R", "COLLAR_L"),
      attackerTriggerR: 0.4,
    })).next;
    const res = tickCutAttempts(s, inputs({
      nowMs: CUT_TIMING.attemptMs,
      attackerRight: gripped("R", "COLLAR_L"),
      attackerTriggerR: 0.9, // attacker re-asserted grip
    }));
    expect(res.events.some((e) => e.kind === "CUT_FAILED" && "defender" in e && e.defender === "R")).toBe(true);
    expect(res.next.right.kind).toBe("IDLE");
  });

  it("mid-attempt second commit is ignored", () => {
    let s = INITIAL_CUT_ATTEMPTS;
    s = tickCutAttempts(s, inputs({
      nowMs: 0,
      leftCommit: { rs: { x: 0, y: 0 } },
      attackerLeft: gripped("L", "SLEEVE_R"),
    })).next;
    const res = tickCutAttempts(s, inputs({
      nowMs: 500,
      leftCommit: { rs: { x: 0, y: 0 } }, // second commit mid-attempt
      attackerLeft: gripped("L", "SLEEVE_R"),
    }));
    expect(res.events).toEqual([]);
    expect(res.next.left.kind).toBe("IN_PROGRESS");
  });

  it("both defender slots can run in parallel", () => {
    let s = INITIAL_CUT_ATTEMPTS;
    s = tickCutAttempts(s, inputs({
      nowMs: 0,
      leftCommit: { rs: { x: -1, y: 0 } },
      rightCommit: { rs: { x: 1, y: 0 } },
      attackerLeft: gripped("L", "SLEEVE_R"),
      attackerRight: gripped("R", "SLEEVE_L"),
    })).next;
    expect(s.left.kind).toBe("IN_PROGRESS");
    expect(s.right.kind).toBe("IN_PROGRESS");
  });
});

File: src/prototype/unity/Assets/BJJSimulator/Tests/EditMode/CutAttemptTest.cs
// NUnit EditMode mirror of src/prototype/web/tests/unit/cut_attempt.test.ts.
// Each [Test] corresponds to one it(...) case from the Stage 1 Vitest suite so
// a regression on either side produces a named, greppable failure.
//
// Reference: docs/design/state_machines_v1.md §4.2 (defender cut-attempt FSM)
//            docs/design/input_system_defense_v1.md §B.4.1 (target picking).
//
// Run from Unity Editor: Window → General → Test Runner → EditMode.

using System.Collections.Generic;
using NUnit.Framework;

namespace BJJSimulator.Tests
{
    // -------------------------------------------------------------------------
    // Helpers (mirrors the gripped() / idle() / inputs() factories in TS).
    // -------------------------------------------------------------------------

    public static class CutAttemptTestHelpers
    {
        public static HandFSM Gripped(HandSide side, GripZone zone) => new HandFSM
        {
            Side            = side,
            State           = HandState.Gripped,
            Target          = zone,
            StateEnteredMs  = 0,
            ReachDurationMs = 0,
            LastParriedZone = GripZone.None,
            LastParriedAtMs = BJJConst.SentinelTimeMs,
        };

        public static HandFSM Idle(HandSide side) => new HandFSM
        {
            Side            = side,
            State           = HandState.Idle,
            Target          = GripZone.None,
            StateEnteredMs  = 0,
            ReachDurationMs = 0,
            LastParriedZone = GripZone.None,
            LastParriedAtMs = BJJConst.SentinelTimeMs,
        };

        // Builds a CutTickInput with sensible defaults; pass overrides as named args.
        public static CutTickInput Inputs(
            long       nowMs            = 0,
            CutCommit? leftCommit       = null,
            CutCommit? rightCommit      = null,
            HandFSM?   attackerLeft     = null,
            HandFSM?   attackerRight    = null,
            float      attackerTriggerL = 0f,
            float      attackerTriggerR = 0f) =>
            new CutTickInput
            {
                NowMs            = nowMs,
                LeftCommit       = leftCommit,
                RightCommit      = rightCommit,
                AttackerLeft     = attackerLeft  ?? Idle(HandSide.L),
                AttackerRight    = attackerRight ?? Idle(HandSide.R),
                AttackerTriggerL = attackerTriggerL,
                AttackerTriggerR = attackerTriggerR,
            };
    }

    // -------------------------------------------------------------------------
    // describe("pickCutTarget (§B.4.1)")
    // -------------------------------------------------------------------------

    [TestFixture]
    public class PickCutTargetTests
    {
        // it("returns null when no attacker hand is GRIPPED")
        [Test]
        public void ReturnsNullWhenNoAttackerHandIsGripped()
        {
            bool ok = CutAttemptOps.PickCutTarget(
                new Vec2 { X = 0f, Y = 0f },
                CutAttemptTestHelpers.Idle(HandSide.L),
                CutAttemptTestHelpers.Idle(HandSide.R),
                out _, out _);

            Assert.IsFalse(ok);
        }

        // it("returns the only GRIPPED hand regardless of RS")
        [Test]
        public void ReturnsTheOnlyGrippedHandRegardlessOfRs()
        {
            bool ok = CutAttemptOps.PickCutTarget(
                new Vec2 { X = 0.5f, Y = 0.5f },
                CutAttemptTestHelpers.Gripped(HandSide.L, GripZone.SleeveR),
                CutAttemptTestHelpers.Idle(HandSide.R),
                out var side, out var zone);

            Assert.IsTrue(ok);
            Assert.AreEqual(HandSide.L,      side);
            Assert.AreEqual(GripZone.SleeveR, zone);
        }

        // it("prefers attacker L when RS points left")
        [Test]
        public void PrefersAttackerLWhenRsPointsLeft()
        {
            bool ok = CutAttemptOps.PickCutTarget(
                new Vec2 { X = -1f, Y = 0f },
                CutAttemptTestHelpers.Gripped(HandSide.L, GripZone.SleeveR),
                CutAttemptTestHelpers.Gripped(HandSide.R, GripZone.SleeveL),
                out var side, out _);

            Assert.IsTrue(ok);
            Assert.AreEqual(HandSide.L, side);
        }

        // it("prefers attacker R when RS points right")
        [Test]
        public void PrefersAttackerRWhenRsPointsRight()
        {
            bool ok = CutAttemptOps.PickCutTarget(
                new Vec2 { X = 1f, Y = 0f },
                CutAttemptTestHelpers.Gripped(HandSide.L, GripZone.SleeveR),
                CutAttemptTestHelpers.Gripped(HandSide.R, GripZone.SleeveL),
                out var side, out _);

            Assert.IsTrue(ok);
            Assert.AreEqual(HandSide.R, side);
        }
    }

    // -------------------------------------------------------------------------
    // describe("tickCutAttempts lifecycle")
    // -------------------------------------------------------------------------

    [TestFixture]
    public class TickCutAttemptsLifecycleTests
    {
        // it("commits start IN_PROGRESS with a target")
        [Test]
        public void CommitsStartInProgressWithATarget()
        {
            var events = new List<CutTickEvent>();
            var next = CutAttemptOps.Tick(
                CutAttempts.Initial,
                CutAttemptTestHelpers.Inputs(
                    nowMs:        0,
                    leftCommit:   new CutCommit { Rs = new Vec2 { X = 0f, Y = 0f } },
                    attackerLeft: CutAttemptTestHelpers.Gripped(HandSide.L, GripZone.SleeveR)),
                events);

            Assert.AreEqual(CutSlotKind.InProgress, next.Left.Kind);
            Assert.IsTrue(events.Exists(e => e.Kind == CutEventKind.CutStarted));
        }

        // it("silent drop when no GRIPPED hand available")
        [Test]
        public void SilentDropWhenNoGrippedHandAvailable()
        {
            var events = new List<CutTickEvent>();
            var next = CutAttemptOps.Tick(
                CutAttempts.Initial,
                CutAttemptTestHelpers.Inputs(
                    leftCommit: new CutCommit { Rs = new Vec2 { X = 1f, Y = 0f } }),
                events);

            Assert.AreEqual(CutSlotKind.Idle, next.Left.Kind);
            Assert.AreEqual(0, events.Count);
        }

        // it("resolution SUCCEEDS when attacker trigger < 0.5 at expiry")
        [Test]
        public void ResolutionSucceedsWhenAttackerTriggerBelowHalfAtExpiry()
        {
            var s = CutAttempts.Initial;
            var events = new List<CutTickEvent>();

            // Tick 1: attacker is gripping hard (0.8) but defender starts the cut.
            s = CutAttemptOps.Tick(
                s,
                CutAttemptTestHelpers.Inputs(
                    nowMs:            0,
                    leftCommit:       new CutCommit { Rs = new Vec2 { X = 0f, Y = 0f } },
                    attackerLeft:     CutAttemptTestHelpers.Gripped(HandSide.L, GripZone.SleeveR),
                    attackerTriggerL: 0.8f),
                events);
            events.Clear();

            // Tick 2: at expiry, attacker has let go (0.3 < 0.5) → cut succeeds.
            s = CutAttemptOps.Tick(
                s,
                CutAttemptTestHelpers.Inputs(
                    nowMs:            CutTiming.Default.AttemptMs,
                    attackerLeft:     CutAttemptTestHelpers.Gripped(HandSide.L, GripZone.SleeveR),
                    attackerTriggerL: 0.3f),
                events);

            Assert.IsTrue(events.Exists(e =>
                e.Kind == CutEventKind.CutSucceeded && e.AttackerSide == HandSide.L));
            Assert.AreEqual(CutSlotKind.Idle, s.Left.Kind);
        }

        // it("resolution FAILS when attacker trigger ≥ 0.5 at expiry")
        [Test]
        public void ResolutionFailsWhenAttackerTriggerAtLeastHalfAtExpiry()
        {
            var s = CutAttempts.Initial;
            var events = new List<CutTickEvent>();

            s = CutAttemptOps.Tick(
                s,
                CutAttemptTestHelpers.Inputs(
                    nowMs:            0,
                    rightCommit:      new CutCommit { Rs = new Vec2 { X = 0f, Y = 0f } },
                    attackerRight:    CutAttemptTestHelpers.Gripped(HandSide.R, GripZone.CollarL),
                    attackerTriggerR: 0.4f),
                events);
            events.Clear();

            // Attacker re-asserted grip during the 1500ms attempt.
            s = CutAttemptOps.Tick(
                s,
                CutAttemptTestHelpers.Inputs(
                    nowMs:            CutTiming.Default.AttemptMs,
                    attackerRight:    CutAttemptTestHelpers.Gripped(HandSide.R, GripZone.CollarL),
                    attackerTriggerR: 0.9f),
                events);

            Assert.IsTrue(events.Exists(e =>
                e.Kind == CutEventKind.CutFailed && e.DefenderSide == HandSide.R));
            Assert.AreEqual(CutSlotKind.Idle, s.Right.Kind);
        }

        // it("mid-attempt second commit is ignored")
        [Test]
        public void MidAttemptSecondCommitIsIgnored()
        {
            var s = CutAttempts.Initial;
            var events = new List<CutTickEvent>();

            s = CutAttemptOps.Tick(
                s,
                CutAttemptTestHelpers.Inputs(
                    nowMs:        0,
                    leftCommit:   new CutCommit { Rs = new Vec2 { X = 0f, Y = 0f } },
                    attackerLeft: CutAttemptTestHelpers.Gripped(HandSide.L, GripZone.SleeveR)),
                events);
            events.Clear();

            // Second commit mid-attempt at t=500 (< 1500 expiry).
            s = CutAttemptOps.Tick(
                s,
                CutAttemptTestHelpers.Inputs(
                    nowMs:        500,
                    leftCommit:   new CutCommit { Rs = new Vec2 { X = 0f, Y = 0f } },
                    attackerLeft: CutAttemptTestHelpers.Gripped(HandSide.L, GripZone.SleeveR)),
                events);

            Assert.AreEqual(0, events.Count);
            Assert.AreEqual(CutSlotKind.InProgress, s.Left.Kind);
        }

        // it("both defender slots can run in parallel")
        [Test]
        public void BothDefenderSlotsCanRunInParallel()
        {
            var s = CutAttempts.Initial;
            var events = new List<CutTickEvent>();

            s = CutAttemptOps.Tick(
                s,
                CutAttemptTestHelpers.Inputs(
                    nowMs:         0,
                    leftCommit:    new CutCommit { Rs = new Vec2 { X = -1f, Y = 0f } },
                    rightCommit:   new CutCommit { Rs = new Vec2 { X =  1f, Y = 0f } },
                    attackerLeft:  CutAttemptTestHelpers.Gripped(HandSide.L, GripZone.SleeveR),
                    attackerRight: CutAttemptTestHelpers.Gripped(HandSide.R, GripZone.SleeveL)),
                events);

            Assert.AreEqual(CutSlotKind.InProgress, s.Left.Kind);
            Assert.AreEqual(CutSlotKind.InProgress, s.Right.Kind);
        }
    }
}
